archive/stat.go

```diff
@@ -0,0 +1,170 @@
+package archive
+
+import (
+	"context"
+	"errors"
+	"sort"
+
+	"github.com/brimsec/zq/pkg/iosrc"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zcode"
+	"github.com/brimsec/zq/zdx"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/brimsec/zq/zqe"
+)
+
+// statReadCloser implements zbuf.ReadCloser
+type statReadCloser struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	ark    *Archive
+	zctx   *resolver.Context
+	recs   chan *zng.Record
+	err    error
+
+	chunkBuilder  *zng.Builder
+	indexBuilders map[string]*zng.Builder
+}
+
+func (s *statReadCloser) Read() (*zng.Record, error) {
+	select {
+	case r, ok := <-s.recs:
+		if !ok {
+			return nil, s.err
+		}
+		return r, nil
+	case <-s.ctx.Done():
+		return nil, s.ctx.Err()
+	}
+}
+
+func (s *statReadCloser) Close() error {
+	s.cancel()
+	return nil
+}
+
+func (s *statReadCloser) chunkRecord(si SpanInfo, zardir iosrc.URI) error {
+	fi, err := iosrc.Stat(ZarDirToLog(zardir))
+	if err != nil {
+		return err
+	}
+
+	if s.chunkBuilder == nil {
+		s.chunkBuilder = zng.NewBuilder(s.zctx.MustLookupTypeRecord([]zng.Column{
+			zng.NewColumn("type", zng.TypeString),
+			zng.NewColumn("log_id", zng.TypeString),
+			zng.NewColumn("start", zng.TypeTime),
+			zng.NewColumn("duration", zng.TypeDuration),
+			zng.NewColumn("size", zng.TypeUint64),
+		}))
+	}
+
+	rec := s.chunkBuilder.Build(
+		zng.EncodeString("chunk"),
+		zng.EncodeString(string(si.LogID)),
+		zng.EncodeTime(si.Span.Ts),
+		zng.EncodeDuration(si.Span.Dur),
+		zng.EncodeUint(uint64(fi.Size())),
+	).Keep()
```

Comment:
Calling Keep does an extra copy of both the record and its bytes. (zng.Builder.Build always starts with a fresh (nil, really) buffer because it calls zcode.Builder.Reset before using its buffer.)

Reply:
Without the call to Keep, I saw test failures due to data races; `Build` returns a pointer to its own `rec zng.Record` field. I could just drop using a zng.Builder entirely - I really just want the loop of building the zcode.Bytes over the columns.

Reply:
Ah, I missed that. Keep the Keep, then.